### DISCUSSION TITLE
Any way to easily execute deploy script and revert to the post deployment step before every test

### DISCUSSION THREAD

**Discussion by wighawag** [wagmi-dev/anvil.js#29] [General]
Hey, great job with this.

I would like to use this somehow with my foundry setup and deploy scripts (forge script written in solidity)

Ideally I would like to be able to execute the said scripts once, save the snapshot and revert to it before every test

Or any other mechanism to keep it fast and efficient. Any example on how to do something like that ?

**Comment by fubhy** [comment]
Hey. Thanks.

We actually talked about this a few days ago. I've started working on a `forge.js` package right next to the `anvil.js` one in this mono repo. I'll share the first version shortly. Same thing there... Simply a thin wrapper around all the forge commands. That way, you'd be able to spawn an anvil instance, run a forge script to e.g. deploy (or whatever) and then run your tests afterwards. Saving snapshots already works via RPC methods (`anvil_snapshot`). The `viem` test client also exposes that method.

You could also build that initialization and snapshotting logic into a custom proxy method (see here: https://github.com/wagmi-dev/anvil.js/blob/main/packages/anvil.js/src/proxy/createProxy.ts#L53-L56).

TL;DR: The proxy has some very minimalistic routing going on for some basic methods (e.g. `http://127.0.0.1:8545/{INSTANCE ID}/logs` to fetch logs from a given instance... Or `/start` and `/stop`. You can basically add your own there with a custom fallback handler. For instance a `/reset` one that would do the `forge script` call & snapshot).

**Reply by fubhy** [reply]
Maybe it'd be helpful to expose a blocking `onStart` and `onStop` (for writing caches, logs, etc.) handler to the `pool` of anvil instances so you can add custom logic for whenever an instance is started.

**Comment by nezouse** [comment]
Hey, any updates on forge.js? Would be great to know, so i don't write it manually just to remove it in a few days.

**Reply by fubhy** [reply]
It's coming next week. I got side tracked by a bunch of other tasks but picking it back up again early next week!

**Reply by fubhy** [reply]
Sorry @nezouse. I dropped the ball on this one. Haven't had the time and my use case for this has since disappeared (temporarily?). I'll post here if that changes again in the future!
